File: Problem05/Problem05.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	s := "abadba"
	fmt.Println(longestPalindromeDP02(s))
}

func longestPalindromeDP02(s string) string {
	n := len(s)
	startIndex := 0
	maxlen := 0
	table := make([][]bool, n)
	for i := range table {
		table[i] = make([]bool, n)
	}

	for g := 0; g < n; g++ {
		for i, j := 0, g; j < n; i, j = i+1, j+1 {
			solve(table, i, j, s)
			if table[i][j] {
				if j-i+1 > maxlen {
					startIndex = i
					maxlen = j - i + 1
				}
			}
		}
	}
	return s[startIndex : startIndex+maxlen]
}

func solve(table [][]bool, i, j int, s string) bool {
	if i == j {
		table[i][j] = true
		return true
	}
	if j-i == 1 {
		if s[i] == s[j] {
			table[i][j] = true
			return true
		} else {
			table[i][j] = false
			return false
		}
	}
	if s[i] == s[j] && table[i+1][j-1] {
		table[i][j] = true
		return true
	} else {
		table[i][j] = false
		return false
	}
}

func longestPalindromeDP(s string) string {
	length := len(s)
	table := make([][]bool, length)

	for i := range s {
		table[i] = make([]bool, length)
		table[i][i] = true
	}

	maxLength := 1
	start := 0
	for i := 3; i <= length; i++ {
		for j := 0; j < length-i+1; j++ {
			k := i + j - 1
			if table[j+1][k-1] && s[j] == s[k] {
				table[j][k] = true
				if i > maxLength {
					start = j
					maxLength = i
				}
			}
		}
	}
	return s[start:maxLength]
}

func longestPalindrome02(s string) string {
	var result strings.Builder
	for i := 0; i < len(s); i++ {
		oddPalindrome := expandFromCenter(s, i, i)
		evenPalindrome := expandFromCenter(s, i, i+1)

		if len(oddPalindrome) > result.Len() {
			result.Reset()
			result.WriteString(oddPalindrome)
		}

		if len(evenPalindrome) > result.Len() {
			result.Reset()
			result.WriteString(evenPalindrome)
		}
	}
	return result.String()
}

func expandFromCenter(s string, left int, right int) string {
	for left >= 0 && right < len(s) && s[left] == s[right] {
		left -= 1
		right += 1
	}

	return s[left+1 : right]
}

func longestPalindrome(s string) string {
	var result strings.Builder

	for i := 0; i < len(s); i++ {
		temp := s[i:]
		tempLen := len(temp)
		for tempLen > 0 {
			if isPalindrome(temp) {
				if len(temp) > result.Len() {
					result.Reset()
					result.WriteString(temp)
				}
				break
			} else {
				tempLen--
				temp = temp[0:tempLen]
			}
		}
	}
	return result.String()
}

func isPalindrome(s string) bool {
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}

File: Problem05/Problem05.ts
function longestPalindromeDP02(s: string): string {
    const n = s.length;
    let startIndex = 0;
    let maxlen = 0;
    const table: boolean[][] = Array.from({ length: n }, () => Array(n).fill(false));

    for (let g = 0; g < n; g++) {
        for (let i = 0, j = g; j < n; i++, j++) {
            solve(table, i, j, s);
            if (table[i][j]) {
                if (j - i + 1 > maxlen) {
                    startIndex = i;
                    maxlen = j - i + 1;
                }
            }
        }
    }
    return s.substring(startIndex, startIndex + maxlen);
}

function solve(table: boolean[][], i: number, j: number, s: string): boolean {
    if (i === j) {
        table[i][j] = true;
        return true;
    }
    if (j - i === 1) {
        return table[i][j] = s[i] === s[j];
    }
    if (s[i] === s[j] && table[i + 1][j - 1]) {
        table[i][j] = true;
        return true;
    } else {
        table[i][j] = false;
        return false;
    }
}

function longestPalindromeDP(s: string): string {
    const length = s.length;
    const table: boolean[][] = Array.from({ length }, () => Array(length).fill(false));

    let maxLength = 1;
    let start = 0;

    for (let i = 0; i < length; i++) {
        table[i][i] = true;
    }

    for (let i = 2; i <= length; i++) {
        for (let j = 0; j < length - i + 1; j++) {
            const k = i + j - 1;
            if (table[j + 1][k - 1] && s[j] === s[k]) {
                table[j][k] = true;
                if (i > maxLength) {
                    start = j;
                    maxLength = i;
                }
            }
        }
    }
    return s.substring(start, start + maxLength);
}

function longestPalindrome02(s: string): string {
    let result = "";

    for (let i = 0; i < s.length; i++) {
        const oddPalindrome = expandFromCenter(s, i, i);
        const evenPalindrome = expandFromCenter(s, i, i + 1);

        if (oddPalindrome.length > result.length) {
            result = oddPalindrome;
        }

        if (evenPalindrome.length > result.length) {
            result = evenPalindrome;
        }
    }
    return result;
}

function expandFromCenter(s: string, left: number, right: number): string {
    while (left >= 0 && right < s.length && s[left] === s[right]) {
        left--;
        right++;
    }
    return s.substring(left + 1, right);
}

function longestPalindrome(s: string): string {
    let result = "";

    for (let i = 0; i < s.length; i++) {
        let temp = s.substring(i);
        let tempLen = temp.length;
        while (tempLen > 0) {
            if (isPalindrome(temp)) {
                if (temp.length > result.length) {
                    result = temp;
                }
                break;
            } else {
                tempLen--;
                temp = temp.substring(0, tempLen);
            }
        }
    }
    return result;
}

function isPalindrome(s: string): boolean {
    for (let i = 0; i < s.length / 2; i++) {
        if (s[i] !== s[s.length - 1 - i]) {
            return false;
        }
    }
    return true;
}

console.log(longestPalindromeDP02("abadba"));